dashboard/test/controllers/api/v1/pd/application/teacher_applications_controller_test.rb

```diff
@@ -94,6 +99,7 @@ class TeacherApplicationsControllerTest < ::ActionController::TestCase
       assert_response :success
     end
 
+    # [MEG] TODO: Consider if these two tests should go into teacher_application_test, avoids sign in
```

Comment:
:-) this can get very subjective. controller tests are nice because they give you end-to-end test coverage for hitting the particular controller method. on the other hand, some of our controller and integration test files take minutes to run. this slows down our ability to iterate quickly. 

when to use a model test:
* start by writing unit tests for any substantial logic in your model
* when controller tests are testing the details of the model logic, consider turning them into model tests

when to use a controller test:
* auth tests
* to test a controller method's behavior end-to-end
* to test any other behavior that you couldn't capture in your model tests

when to use an integration test:
* when you need to make multiple api calls, follow redirects, or use other features (see dashboard/test/integration for examples)

considering all that... the two tests below seem good to me to leave as controller tests, because they have the benefit of making it super clear that these important things will happen as the result of a single create (i.e. they are testing end-to-end). if you were going to go and add more detailed tests for what goes on in on_successful_create, then it might be a good time to start adding some model tests instead.

Reply:
Ah, this is gold. Thank you!